### PR TITLE
Add BroadcastTo_001 

### PR DESCRIPTION
**Pull request by Seunghui98**
This adds a recipe file for builtin BroadcastTo.

ONE-DCO-1.0-Signed-off-by: SeungHui Lee <shsh1004.lee@samsung.com>

---
Related to Issue: https://github.com/Samsung/ONE/issues/10784
Draft PR: [#10737](https://github.com/Samsung/ONE/pull/11701)

### CODE REVIEW

**Comment by Seunghui98** [comment]
since `version` can be both 2 and 3 ** [reference](https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311), recipe files was created to test both versions.

**Comment by seanshpark** [comment]
plz split for "Add" and "Change"

File: res/TensorFlowLiteRecipes/BroadcastTo_000/test.recipe [hunk 1]
```diff
@@ -15,11 +15,11 @@ operand {
   shape { dim: 1 dim: 2 dim: 3 }
 }
 operation {
-  type: "Custom"
+  type: "BroadcastTo"
   input: "bc_input"
   input: "bc_shape"
   output: "bc_ofm"
-  custom_code: "BroadcastTo"
+  version: 2
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_000/test.recipe:22]: Q) why is this changed?

Comment (Seunghui98) [reply res/TensorFlowLiteRecipes/BroadcastTo_000/test.recipe:22]: @seanshpark 
> Q) why is this changed?

`BroadcastTo` was newly introduced as a built-in op, so I modified it to fit the existing broadcastTo_000 recipe file. 
Would it be better to make `BroadcastTo_001` and `BroadcastTo_002` recipe files separately for Builtin op test use?

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_000/test.recipe:22]: > Would it be better to make BroadcastTo_001 and BroadcastTo_002 recipe files separately for Builtin op test use?

Yes

Comment (Seunghui98) [reply res/TensorFlowLiteRecipes/BroadcastTo_000/test.recipe]: @seanshpark 

Thank you for your advice !! 
I applied comments. :)

**Review by mhs4670go — DISMISSED** [review]
LGTM

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  name: "bc_ofm"
+  type: FLOAT32
+  shape { dim: 1 dim: 2 dim: 3 }
+}
+operation {
+  type: "BroadcastTo"
+  input: "bc_input"
+  input: "bc_shape"
+  output: "bc_ofm"
+  version: 2
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:22]: why is this version added?

File: res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  name: "bc_ofm"
+  type: FLOAT32
+  shape { dim: 1 dim: 2 dim: 3 }
+}
+operation {
+  type: "BroadcastTo"
+  input: "bc_input"
+  input: "bc_shape"
+  output: "bc_ofm"
+  version: 3
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: why is this 3?

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe (hunk 2 above)
Comment (Seunghui98) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:22]: @seanshpark 
This is because `luci-value-test` failure occurs when the value of the `version` is 1(default value). ->[error message](https://github.com/Samsung/ONE/pull/11701#issuecomment-1767390375).


In tensorflow, the version of the `builtin` op of `broadcastTo` is shown as 2 or 3, 
so I created recipe files to test both of these. :) ** [reference](https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311)

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:22]: OK, thanks. Please add a comment that we have to use version 2 for this Op.

File: res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe (hunk 3 above)
Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: Do we have any conditional branch or test with version is 2 or 3 ?
if not, I don't think we have to add version 3.

Comment (mhs4670go) [reply res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: Or, we can add the check if the version is 2 or 3 in the `validate` function of `compiler/luci/import/src/Nodes/CircleBroadcastTo.cpp`.

```c++
bool CircleBroadcastToGraphBuilder::validate(const ValidateArgs &args) const
{
  auto const bt_version = args.reader.opcodes().at(args.op.opcode_index)->version();
  // The version one of broadcast to op won't be not supported
  // https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311
  if (bt_version != 2 && bt_version != 3)
    return false;

  return GraphBuilder::validate(args, 2);
}
```

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: As I know, for all other Ops we didn't do any version checking: min, max values.
I don't know our IR has version... -_-;
For now, let's skip this version check in import.

Comment (mhs4670go) [reply res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: Okay. Then how about use version 3? I'm not sure but latest version might be good..?

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_002/test.recipe:22]: > ... but latest version might be good..?

Found that circle IR doesn't have version thus doesn't import version per Op.
The version may effect tflite interpreter for value check.
Any version is OK to me with working value check.

**Review by mhs4670go — APPROVED** [review]
LGTM

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# We have to use version 3 for BroadcastTo Op.
```
Comment (Seunghui98) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:1]: @seanshpark 

To mark this, I wrote annotations on the recipe file. 
Should I mark it on other parts? :)

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# We have to use the latest version of 3 for BroadcastTo Op.
+# https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311
```
Comment (Seunghui98) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:2]: @seanshpark 

> OK, thanks. Please add a comment that we have to use version 2 for this Op.

To mark this, I wrote annotations on the recipe file. 
Should I mark it on other parts? :)

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:2]: > Should I mark it on other parts? :)

Sorry but I don't understand your question.

> We have to use the latest version of 3 for BroadcastTo Op.

Why is it so? why not 2? URL you've added doesn't seem to explain this.

My comment about "we have to use version 2" is that this Op is the first Op that requires version > 1. I wanted some explanation what you understand about this.

Comment (mhs4670go) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:2]: I've found related PR on tensorflow. (https://github.com/tensorflow/tensorflow/pull/46224/files)

```cxx
// The version one of broadcast to op won't be not supported since the
    // version one was rollbacked and the builtin op code number has been
    // changed because of builtin op code shortage problem.
    // Quantized broadcast_to is version 3
    case BuiltinOperator_BROADCAST_TO:
      if (op_sig.inputs.at(0).type == kTfLiteInt8 ||
          op_sig.inputs.at(0).type == kTfLiteInt16) {
        return 3;
      }
      return 2;
```

We are using tflite kernels for float values. So, the version should be 2.

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# BroadcastTo Op supports op version 2 and 3.
+# BroadcastTo of Float type is version 2 and Quantized BroadcastTo is version 3.
+# We are using tflite kernels for float values. so, the version should be 2.
+# https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311
+# https://github.com/tensorflow/tensorflow/pull/46224/files
```
Comment (Seunghui98) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:5]: @seanshpark 

based on the comments you gave, I commented in detail so that users can understand the version. 
Are the comments i added okay? :)

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# BroadcastTo Op supports op version 2 and 3.
+# BroadcastTo of Float type is version 2 and Quantized BroadcastTo is version 3.
+# We are using tflite kernels for float values. so, the version should be 2.
+# https://github.com/tensorflow/tensorflow/blob/ce2d218fc88c20b1d81abb017eabd436779865e4/tensorflow/lite/kernels/register.cc#L306-L311
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:4]: I think this can be removed. this line is too long and 46224 explains about the version.

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe (hunk 6 above)
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:5]:
```suggestion
# refer https://github.com/tensorflow/tensorflow/pull/46224/files
```

Comment (Seunghui98) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe]: Thank you a lot !!

**Comment by seanshpark** [comment]
```
# BroadcastTo of Float type is version 2 and Quantized BroadcastTo is version 3.
# We are using tflite kernels for float values. so, the version should be 2.
```
looks fine. I didn't leave any comments. why was it removed?

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe [hunk 8]
```diff
@@ -0,0 +1,28 @@
+# BroadcastTo of Float type is version 2 and Quantized BroadcastTo is version 3.
+# We are using tflite kernels for float values. so, the version should be 2.
+# refer https://github.com/tensorflow/tensorflow/pull/46224/files
```
Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:3]: this comment is for 1~3

Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:3]: this comment is for 3

Comment (seanshpark) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:3]: this comment is for 3

**Comment by seanshpark** [comment]
UI doesn't seem to distinguish much for comment for 1~3 vs 3 only.
there is `Comment on lines +1 to +3` but it's small and not noticeable...

File: res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe (hunk 8 above)
Comment (Seunghui98) [inline res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:3]: @seanshpark 

I misunderstood to remove code lines 1 to 4. 😅 
I modified it to 1~3 lines again, is it okay?

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:3]: > I modified it to 1~3 lines again, is it okay?

I don't quite understand your question.

Comment (Seunghui98) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:4]: @seanshpark 
You gave comments to one line, but I misunderstood to delete the 1 to 3 code lines because of UI. 😅
I modified it again to reflect the comments you gave me, is it okay?

Comment (seanshpark) [reply res/TensorFlowLiteRecipes/BroadcastTo_001/test.recipe:4]: Yes, looks OK. We have some trouble in communication. I hope it gets better next time.

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM
